Add predicate pushdown (updated)

mike-luabase opened the pull request:
This PR enhances query performance by filtering data at the metadata level, reducing the amount of data read during scans. It currently only addresses queries containing a single Iceberg table.

### **Key Changes**

- **Extended `IcebergManifestEntry`:**
  - Added `lower_bounds` and `upper_bounds` maps to store column statistics.
  
- **Utility Function:**
  - Implemented `IcebergUtils::GetFullPath` to resolve file paths accurately.

- **Metadata Retrieval:**
  - Added `GetEntries` template in `IcebergTable` to fetch relevant manifest entries, excluding deleted ones.

- **Predicate Evaluation:**
  - Created `EvaluatePredicateAgainstStatistics` to assess if data files satisfy query predicates based on their statistics.
  - For each predicate:
    - Identifies the column involved.
    - Checks if the column has defined lower and upper bounds.
    - Based on the comparison type (e.g., =, >, <), determines if the predicate can be satisfied given the file's bounds.
    - If any predicate fails, the file is excluded from the scan.

- **Scan Expression Modification:**
  - Updated `MakeScanExpression` to filter `data_file_entries` using predicates before scanning.

- **Binding Function Enhancements:**
  - Enhanced `IcebergScanBindReplace` with additional logging and prepared data files based on predicate results.

This implementation optimizes Iceberg table scans by leveraging metadata for early data filtering, significantly improving query efficiency and resource usage.

Mytherin commented on src/iceberg_functions/iceberg_scan.cpp:
```diff
@@ -19,321 +19,659 @@
-	select_statement->node = std::move(select_node);
-	return make_uniq<SubqueryRef>(std::move(select_statement), "iceberg_scan");
+    vector<Value> structs;
+    for (const auto &file : data_file_values) {
```
The majority of lines changed here are still formatting changes - could you [format using our clang-format config](https://github.com/duckdb/duckdb/blob/main/.clang-format)?

mike-luabase replied:
@Mytherin fixed the formatting, sorry about that!

samansmink commented:
Also @mike-luabase, you don't need to reopen the PR every time, this makes it harder to review actually since we lose a clear view on the review comments that were made before.

If you want to overwrite commits you can just force push to your feature branch

mike-luabase commented:
> Also @mike-luabase, you don't need to reopen the PR every time, this makes it harder to review actually since we lose a clear view on the review comments that were made before.

Yes, understood, won't do that again.

mike-luabase commented:
This test is failing, looking into it

```
SELECT count(*) FROM ICEBERG_SCAN('data/iceberg/generated_spec2_0_001/pyspark_iceberg_table');
Lower bounds is null
Upper bounds is null
Binder Error: Table "iceberg_scan_data" does not have a column named "filename"
```

mike-luabase commented:
@Mytherin test failure should be fixed now

mike-luabase commented:
@samansmink @Mytherin can we trigger the run again? This error doesn't seem related to my changes:

```
#7 [ 3/14] RUN apt-get install -y -qq software-properties-common
#7 0.232 E: Unable to locate package software-properties-common
#7 ERROR: process "/bin/sh -c apt-get install -y -qq software-properties-common" did not complete successfully: exit code: 100
------
 > [ 3/14] RUN apt-get install -y -qq software-properties-common:
0.232 E: Unable to locate package software-properties-common
------
Dockerfile:9
--------------------
   7 |     # Setup the basic necessities
   8 |     RUN apt-get update -y -qq
   9 | >>> RUN apt-get install -y -qq software-properties-common
  10 |     RUN apt-get install -y -qq --fix-missing ninja-build make gcc-multilib g++-multilib libssl-dev wget openjdk-8-jdk zip maven unixodbc-dev libc6-dev-i386 lib32readline6-dev libssl-dev libcurl4-gnutls-dev libexpat1-dev gettext unzip build-essential checkinstall libffi-dev curl libz-dev openssh-client pkg-config autoconf
  11 |     RUN apt-get install -y -qq ccache
--------------------
ERROR: failed to solve: process "/bin/sh -c apt-get install -y -qq software-properties-common" did not complete successfully: exit code: 100
```

catkins commented:
@mike-luabase 💚 I'm so excited to see movement on this! I'll be able to delete a bunch of manual partition pruning + read_parquet code in an app of mine!

mike-luabase commented:
@samansmink anything I can do to help here?

samansmink commented:
If you could re-add a PR description and address @Mytherin's comment from one of your previous PRs https://github.com/duckdb/duckdb_iceberg/pull/75#discussion_r1800017854, I will take a more detailed look to review this.

Please be considerate that reviews take a lot of time, especially reviews from outside contributors that are touching complicated code. To get your PRs through as quick as possible, I recommend you to:
- double check your ideas for a PR with a core contributor of DuckDB through either the DuckDB discord channel, a github issue or a github discussion. See (https://github.com/duckdb/duckdb/blob/main/CONTRIBUTING.md)
- Make sure that your PRs are easy to review
    - no big PRs
    - don't add big format changes to your PRs
    - write a clear description of what you are adding and why you chose the path you chose

mike-luabase commented:
@samansmink I added the description and fixed the formatting!

I understand this PR is on the larger side, but I think it's as small as it could be to get predicate pushdown working.

I'm more than happy to discuss the PR! I'll drop a note in the Discord.

Mytherin reviewed (commented):
Thanks for reducing the changeset - this is easier to review. I still wonder about the changes to the Avro files. In addition - there's no tests included with the changeset. Could you add tests that stress test the predicate pushdown in various ways (e.g. testing out different predicate types, different data types, etc)?

mike-luabase commented:
on it!

mike-luabase commented on src/include/avro_codegen/iceberg_manifest_entry_partial.hpp:
```diff
@@ -1,124 +1,380 @@
 
+/* This code was generated by avrogencpp 1.13.0-SNAPSHOT. Do not edit.*/
+
+#ifndef MANIFEST_ENTRY_HH_2043678367_H
```
@Mytherin the existing types didn't have a way to access `lower_bounds` and `upper_bounds`, which we need for pushdown.

mike-luabase commented:
@Mytherin I added tests and updated the data generation for predicate pushdown. Let me know if that works!

mike-luabase commented:
I see files from the `/data/iceberg` have somehow made it back into the PR. I'm trying to remove them, but not having much luck.

samansmink commented:
As mentioned before, @mike-luabase why is this PR removing the checked in test data?

peterboncz commented on scripts/test_data_generator/generate_iceberg.py:
```diff
@@ -1,4 +1,4 @@
-#!/usr/bin/python3
+#!/Users/mritchie712/opt/anaconda3/bin/python
```
this will break on other computers. In this file there are some useful changes (partitioning) but also spurious changes likes semicolons and spacing.. It would be nice to remove the spurious changes to make the diff as small as possible

peterboncz commented on src/common/iceberg.cpp:
```diff
@@ -80,8 +80,8 @@ vector<IcebergManifestEntry> IcebergTable::ReadManifestEntries(const string &pat
 		auto schema = avro::compileJsonSchemaFromString(MANIFEST_ENTRY_SCHEMA);
-		avro::DataFileReader<c::manifest_entry> dfr(std::move(stream), schema);
-		c::manifest_entry manifest_entry;
+		avro::DataFileReader<manifest_entry> dfr(std::move(stream), schema);
```
these changes are spurious and hence this file change could be removed from the PR

peterboncz reviewed (commented):
Thanks for the efforts, Mike. Was looking through it; but I would not call it a review and in any case I think DuckDB Labs folks need to do that. I marked a number of diffs in the PR that I think could be removed: the deletion of the generated files (as already remarked by Sam) and some smaller things.

nammaste6 commented on src/iceberg_functions/iceberg_scan.cpp:
```diff
@@ -36,6 +38,34 @@ struct IcebergScanGlobalTableFunctionState : public GlobalTableFunctionState {
+
+// === Helper function to recursively extract comparison predicates from expressions ===
+static void ExtractPredicates(ParsedExpression &expr, vector<unique_ptr<ParsedExpression>> &predicates) {
+	if (expr.type == ExpressionType::CONJUNCTION_AND) {
+		auto &conj = (ConjunctionExpression &)expr;
+		// Access children instead of left and right
+		if (conj.children.size() >= 2) {
+			ExtractPredicates(*conj.children[0], predicates);
+			ExtractPredicates(*conj.children[1], predicates);
+		}
+	} else if (expr.type == ExpressionType::COMPARE_EQUAL || expr.type == ExpressionType::COMPARE_GREATERTHAN ||
+	           expr.type == ExpressionType::COMPARE_GREATERTHANOREQUALTO ||
+	           expr.type == ExpressionType::COMPARE_LESSTHAN ||
+	           expr.type == ExpressionType::COMPARE_LESSTHANOREQUALTO) {
+		// Clone the expression and add to predicates
+		predicates.emplace_back(expr.Copy());
+	}
+	// Add more conditions here if you want to handle OR or other expressions
```
@mike-luabase I’d like to suggest adding between and fixing a bug in this part.
Would it be okay if I add the commit directly? Or would it be better for me to submit a separate PR?

mike-luabase replied:
@nammaste6 please check https://github.com/duckdb/duckdb-iceberg/pull/101

they are looking to implement this upstream of the extension in MultiFileReader

samansmink commented:
Hey @mike-luabase, the multifilereader refactor is now merged (https://github.com/duckdb/duckdb-iceberg/pull/101). We will look into implementing predicate pushdown very soon using: [this](https://github.com/duckdb/duckdb-iceberg/blob/6f02e6237199c2935c77aede9a23ba4bdc96ab5f/src/iceberg_functions/iceberg_multi_file_reader.cpp#L41) api call

mike-luabase commented:
@samansmink great news, thanks for letting me know!
